File: e2e-cli/src/main/java/cli/Main.java
package cli;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.segment.analytics.Analytics;
import com.segment.analytics.Callback;
import com.segment.analytics.messages.*;

import java.lang.reflect.Type;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

public class Main {

    private static final Gson gson = new Gson();
    private static final int DEFAULT_MAX_RETRIES = 1000;

    public static void main(String[] args) {
        Map<String, Object> output;
        try {
            output = run(args);
        } catch (Exception e) {
            output = errorOutput(e.getMessage());
        }
        System.out.println(gson.toJson(output));
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> run(String[] args) throws Exception {
        int inputIndex = indexOf(args, "--input");
        if (inputIndex == -1 || inputIndex + 1 >= args.length) {
            throw new IllegalArgumentException("Missing required --input argument");
        }

        String inputJson = args[inputIndex + 1];
        Type mapType = new TypeToken<Map<String, Object>>() {}.getType();
        Map<String, Object> input = gson.fromJson(inputJson, mapType);

        String writeKey = (String) input.get("writeKey");
        String apiHost = (String) input.get("apiHost");
        Map<String, Object> config = (Map<String, Object>) input.getOrDefault("config", Collections.emptyMap());
        List<Map<String, Object>> sequences = (List<Map<String, Object>>) input.get("sequences");

        int flushAt = intVal(config, "flushAt", 20);
        long flushIntervalMs = longVal(config, "flushInterval", 10000L);
        int maxRetries = intVal(config, "maxRetries", DEFAULT_MAX_RETRIES);

        AtomicBoolean hasError = new AtomicBoolean(false);
        AtomicReference<String> errorMessage = new AtomicReference<>();

        Analytics analytics = Analytics.builder(writeKey)
                .endpoint(apiHost)
                .flushQueueSize(flushAt)
                .flushInterval(Math.max(flushIntervalMs, 1000L), TimeUnit.MILLISECONDS)
                .retries(maxRetries)
                .callback(new Callback() {
                    @Override
                    public void success(Message message) {
                    }

                    @Override
                    public void failure(Message message, Throwable throwable) {
                        hasError.set(true);
                        errorMessage.set(throwable != null ? throwable.getMessage() : "unknown error");
                    }
                })
                .build();

        for (Map<String, Object> seq : sequences) {
            long delayMs = longVal(seq, "delayMs", 0L);
            if (delayMs > 0) {
                Thread.sleep(delayMs);
            }
            List<Map<String, Object>> events = (List<Map<String, Object>>) seq.get("events");
            if (events != null) {
                for (Map<String, Object> event : events) {
                    sendEvent(analytics, event);
                }
            }
        }

        analytics.flush();
        analytics.shutdown();

        if (hasError.get()) {
            return errorOutput(errorMessage.get());
        }

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("success", true);
        result.put("sentBatches", 1);
        return result;
    }

    @SuppressWarnings("unchecked")
    private static void sendEvent(Analytics analytics, Map<String, Object> event) {
        String type = (String) event.get("type");
        if (type == null) {
            throw new IllegalArgumentException("Event missing 'type' field");
        }

        String userId = strVal(event, "userId", "");
        String anonymousId = (String) event.get("anonymousId");
        String messageId = (String) event.get("messageId");
        String timestamp = (String) event.get("timestamp");
        Map<String, Object> traits = (Map<String, Object>) event.getOrDefault("traits", Collections.emptyMap());
        Map<String, Object> properties = (Map<String, Object>) event.getOrDefault("properties", Collections.emptyMap());
        String eventName = (String) event.get("event");
        String name = (String) event.get("name");
        String groupId = (String) event.get("groupId");
        String previousId = (String) event.get("previousId");
        Map<String, Object> context = (Map<String, Object>) event.get("context");
        Map<String, Object> integrations = (Map<String, Object>) event.get("integrations");

        MessageBuilder<?, ?> builder;
        switch (type) {
            case "identify":
                builder = IdentifyMessage.builder().traits(traits);
                break;
            case "track":
                builder = TrackMessage.builder(eventName != null ? eventName : "Unknown Event").properties(properties);
                break;
            case "page":
                builder = PageMessage.builder(name != null ? name : "Unknown Page").properties(properties);
                break;
            case "screen":
                builder = ScreenMessage.builder(name != null ? name : "Unknown Screen").properties(properties);
                break;
            case "alias":
                builder = AliasMessage.builder(previousId != null ? previousId : "");
                break;
            case "group":
                builder = GroupMessage.builder(groupId != null ? groupId : "").traits(traits);
                break;
            default:
                throw new IllegalArgumentException("Unknown event type: " + type);
        }

        if (!userId.isEmpty()) {
            builder.userId(userId);
        }
        if (anonymousId != null) {
            builder.anonymousId(anonymousId);
        }
        if (messageId != null) {
            builder.messageId(messageId);
        }
        if (timestamp != null) {
            builder.timestamp(Date.from(Instant.parse(timestamp)));
        }
        if (context != null) {
            builder.context(context);
        }
        if (integrations != null) {
            for (Map.Entry<String, Object> entry : integrations.entrySet()) {
                Object value = entry.getValue();
                if (value instanceof Boolean) {
                    builder.enableIntegration(entry.getKey(), (Boolean) value);
                } else if (value instanceof Map) {
                    builder.integrationOptions(entry.getKey(), (Map<String, ?>) value);
                }
            }
        }

        analytics.enqueue(builder);
    }

    private static Map<String, Object> errorOutput(String error) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("success", false);
        result.put("error", error);
        result.put("sentBatches", 0);
        return result;
    }

    private static int indexOf(String[] arr, String target) {
        for (int i = 0; i < arr.length; i++) {
            if (target.equals(arr[i])) return i;
        }
        return -1;
    }

    private static int intVal(Map<String, Object> map, String key, int defaultVal) {
        Object v = map.get(key);
        if (v instanceof Number) return ((Number) v).intValue();
        return defaultVal;
    }

    private static long longVal(Map<String, Object> map, String key, long defaultVal) {
        Object v = map.get(key);
        if (v instanceof Number) return ((Number) v).longValue();
        return defaultVal;
    }

    private static String strVal(Map<String, Object> map, String key, String defaultVal) {
        Object v = map.get(key);
        if (v instanceof String) return (String) v;
        return defaultVal;
    }
}

File: e2e-cli/src/main/kotlin/cli/Main.kt
package cli

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.segment.analytics.Analytics
import com.segment.analytics.Callback
import com.segment.analytics.messages.*
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

data class CLIOutput(
    val success: Boolean,
    val error: String? = null,
    val sentBatches: Int = 0
)

data class CLIConfig(
    val flushAt: Int? = null,
    val flushInterval: Long? = null,
    val maxRetries: Int? = null,
    val timeout: Int? = null
)

data class EventSequence(
    val delayMs: Long = 0,
    val events: List<Map<String, Any>>
)

data class CLIInput(
    val writeKey: String,
    val apiHost: String,
    val sequences: List<EventSequence>,
    val config: CLIConfig? = null
)

private val gson = Gson()

fun main(args: Array<String>) {
    var output = CLIOutput(success = false, error = "Unknown error")

    try {
        // Parse --input argument
        val inputIndex = args.indexOf("--input")
        if (inputIndex == -1 || inputIndex + 1 >= args.size) {
            throw IllegalArgumentException("Missing required --input argument")
        }

        val inputJson = args[inputIndex + 1]
        val input = gson.fromJson(inputJson, CLIInput::class.java)

        val flushAt = input.config?.flushAt ?: 20
        val flushIntervalMs = input.config?.flushInterval ?: 10000L

        val flushLatch = CountDownLatch(1)
        val hasError = AtomicBoolean(false)
        var errorMessage: String? = null

        val analytics = Analytics.builder(input.writeKey)
            .endpoint(input.apiHost)
            .flushQueueSize(flushAt)
            .flushInterval(maxOf(flushIntervalMs, 1000L), TimeUnit.MILLISECONDS)
            .callback(object : Callback {
                override fun success(message: Message?) {
                    // Event sent successfully
                }

                override fun failure(message: Message?, throwable: Throwable?) {
                    hasError.set(true)
                    errorMessage = throwable?.message
                }
            })
            .build()

        // Process event sequences
        for (seq in input.sequences) {
            if (seq.delayMs > 0) {
                Thread.sleep(seq.delayMs)
            }

            for (event in seq.events) {
                sendEvent(analytics, event)
            }
        }

        // Flush and shutdown
        analytics.flush()
        analytics.shutdown()

        output = if (hasError.get()) {
            CLIOutput(success = false, error = errorMessage, sentBatches = 0)
        } else {
            CLIOutput(success = true, sentBatches = 1)
        }

    } catch (e: Exception) {
        output = CLIOutput(success = false, error = e.message ?: e.toString())
    }

    println(gson.toJson(output))
}

fun sendEvent(analytics: Analytics, event: Map<String, Any>) {
    val type = event["type"] as? String
        ?: throw IllegalArgumentException("Event missing 'type' field")

    val userId = event["userId"] as? String ?: ""
    val anonymousId = event["anonymousId"] as? String
    val messageId = event["messageId"] as? String
    val timestamp = event["timestamp"] as? String
    @Suppress("UNCHECKED_CAST")
    val traits = event["traits"] as? Map<String, Any> ?: emptyMap()
    @Suppress("UNCHECKED_CAST")
    val properties = event["properties"] as? Map<String, Any> ?: emptyMap()
    val eventName = event["event"] as? String
    val name = event["name"] as? String
    val category = event["category"] as? String
    val groupId = event["groupId"] as? String
    val previousId = event["previousId"] as? String
    @Suppress("UNCHECKED_CAST")
    val context = event["context"] as? Map<String, Any>
    @Suppress("UNCHECKED_CAST")
    val integrations = event["integrations"] as? Map<String, Any>

    val messageBuilder: MessageBuilder<*, *> = when (type) {
        "identify" -> {
            IdentifyMessage.builder().apply {
                traits(traits)
            }
        }
        "track" -> {
            TrackMessage.builder(eventName ?: "Unknown Event").apply {
                properties(properties)
            }
        }
        "page" -> {
            PageMessage.builder(name ?: "Unknown Page").apply {
                properties(properties)
            }
        }
        "screen" -> {
            ScreenMessage.builder(name ?: "Unknown Screen").apply {
                properties(properties)
            }
        }
        "alias" -> {
            AliasMessage.builder(previousId ?: "")
        }
        "group" -> {
            GroupMessage.builder(groupId ?: "").apply {
                traits(traits)
            }
        }
        else -> throw IllegalArgumentException("Unknown event type: $type")
    }

    if (userId.isNotEmpty()) {
        messageBuilder.userId(userId)
    }
    if (anonymousId != null) {
        messageBuilder.anonymousId(anonymousId)
    }
    if (messageId != null) {
        messageBuilder.messageId(messageId)
    }
    if (timestamp != null) {
        messageBuilder.timestamp(parseTimestamp(timestamp))
    }
    if (context != null) {
        messageBuilder.context(context)
    }
    if (integrations != null) {
        for ((key, value) in integrations) {
            when (value) {
                is Boolean -> messageBuilder.enableIntegration(key, value)
                is Map<*, *> -> @Suppress("UNCHECKED_CAST")
                    messageBuilder.integrationOptions(key, value as Map<String, Any>)
            }
        }
    }

    analytics.enqueue(messageBuilder)
}

private fun parseTimestamp(iso: String): Date {
    val format = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US)
    format.timeZone = TimeZone.getTimeZone("UTC")
    return try {
        format.parse(iso)!!
    } catch (_: Exception) {
        // Fallback: try without millis
        val fallback = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US)
        fallback.timeZone = TimeZone.getTimeZone("UTC")
        fallback.parse(iso)!!
    }
}
